Run aXe on a few pages during CI

toolness opened the pull request:
Fixes #326.

This runs [axe-core](https://github.com/dequelabs/axe-core) on a few pages of the standards during CI and `npm test` runs. In future PRs we can add support for running it on more pages.

A successful run looks something like this:

> ![2017-07-12_15-43-34](https://user-images.githubusercontent.com/124687/28136780-fc177cc0-6718-11e7-8803-750b651de81f.png)

It requires that Chrome/Chromium v59 or higher be installed on the host system. I decided on [headless Chrome](https://developers.google.com/web/updates/2017/04/headless-chrome) because:

* It's maintained by the Google Chrome team, so it's likely to be very well-maintained going forward. Possibly even moreso than Phantom; I vaguely recall seeing a Tweet from Ariya Hidayat a while back saying that Phantom was going to be end-of-life'd since Chromium is headless is officially supported now, but I could be totally wrong on that.
* I haven't done any formal measurements, but headless Chrome seems _really_ fast compared to Phantom.
* In the vast majority of cases, it's just one less thing for developers to install--every developer I know has Chrome installed on their system, and our use of `chrome-launcher` ensures that it will be found, regardless of host OS.
* I noticed we have a (kind of old) WIP PR to add support for Google Lighthouse in #210, so this PR might make continuing work on that easier too,
* Chrome is a real browser, unlike PhantomJS, so it's likely that any errors can be reproduced and debugged using Chrome running in non-headless (headed?) mode.

To do:
- [x] Ensure this works on CircleCI.
- [x] Ensure we're running aXe on the actual front page and not e.g. a "network error" page.
- [x] Ensure that [`docker-uswds`](https://github.com/toolness/docker-uswds) can run this too.
- [x] Update README to indicate that developers should have Chrome/Chromium installed on their systems for tests to run properly.
- [x] Modify `crawl.js` to use `static-server.js` to prevent duplication of code.

toolness commented:
Hmm, so this is a fair bit of code, as the [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/) is fairly low-level. There might be a higher-level wrapper available that could reduce the size of our code, though I'm not sure if it'd necessarily meet our particular needs or not. Even axe-core's own PhantomJS wrapper didn't meet all the needs for the [a11y-metrics](https://github.com/18F/a11y-metrics) dashboard so I had to write my own.

msecret commented on config/static-server.js:
```diff
@@ -0,0 +1,30 @@
+}
+
+module.exports = () => {
+  return new Promise((resolve, reject) => {
```
Does this need a rejection handler somewhere?

toolness replied:
Ah good CATCH, will add!

msecret commented on config/run-axe.js:
```diff
@@ -0,0 +1,152 @@
+  const chromeHost = chrome.host || 'localhost';
+  console.log(`Static file server is listening at ${server.url}.`);
+  console.log(`Chrome is debuggable on http://${chromeHost}:${chrome.port}.`);
+  console.log(`Running aXe on:`);
```
Was there more text that was supposed to go here?

toolness replied:
Nah, it's just the heading text that appears before each individual page that's visited. The final output looks like this:

![2017-07-12_15-43-34](https://user-images.githubusercontent.com/124687/28136780-fc177cc0-6718-11e7-8803-750b651de81f.png)

msecret commented on config/run-axe.js:
```diff
@@ -0,0 +1,152 @@
+  console.log(`Running aXe on:`);
+
+  CDP({
+    host: chrome.host,
```
Above, it sounds like `chrome.host` may not exist. Should this be `chromeHost`? Why wouldn't `chrome.host` not exist anyway?

toolness replied:
Oh, it will just be `undefined` if it doesn't exit, which CDP just then defaults to localhost--but I agree, I think it'd be less confusing / more explicit if we just passed `chromeHost` here.

I was hoping that the API for `chrome` would set `host` to whatever the ultimate "resolved" value was, whether that was `localhost` (in the case where Chrome is launched locally) or something explicit (in the case where we connect to a remote Chrome in e.g. a different Docker container).  However, that wasn't the case--in the case where Chrome is launched locally, `host` isn't set at all, so I had to manually default it to `localhost` via that `chromeHost` variable. It's a bit annoying because the APIs that connect all these different Chrome packages isn't quite as consistent as I'd hoped, and this is a workaround for that.

msecret commented on config/run-axe.js:
```diff
@@ -0,0 +1,152 @@
+      } else {
+        const page = pagesLeft.pop();
+        const url = `${server.url}${page}`;
+        process.stdout.write(`  ${page} `);
```
What's this doing? Why isn't it a simple `console.log`?

toolness replied:
Oh, I wanted it to look cool like all the fancy test runners!

![2017-07-12_15-43-34](https://user-images.githubusercontent.com/124687/28136780-fc177cc0-6718-11e7-8803-750b651de81f.png)

msecret commented on config/run-axe.js:
```diff
@@ -0,0 +1,152 @@
+      Page.enable(),
+      Network.enable(),
+    ]).then(() => {
+      Network.responseReceived(({response}) => {
```
Is this when any page resource is received? Or just the main HTML document?

toolness replied:
I believe it's when *any* page resource is received.  So this effectively overlaps a bit with our crawler from #321.  It was easier to do it that way than it was to check the URL and see if it matched the page we were requesting, but I could add such a conditional if you think it'd be better that way.

msecret replied:
Seems OK to have this extra check

toolness replied:
Hmm, so I thought about it a bit, and realized that maybe it's actually good that it's counting any sub-resource failure as a 404 too.  For example, if it's 404'ing on CSS or JS, then that changes the accessibility profile of the page--calculated color contrast ratios may be incorrect, or certain ARIA attributes may not be present on the evaluated page.

Another thing is that the crawler from #321 isn't actually 100% accurate--the library we're using uses regular expressions to "scrape" resources for URLs, so it's actually possible that it might miss sub-resources that are noticed by actually loading a page in Chrome. So it might still be good to hard fail on sub-resource failures for the sake of redundancy, too.

What do you think? At the very least, I think we should log warnings on sub-resource fails.

msecret replied:
Yea I agree that it's good to consider sub-resource 404s as failures.

msecret commented on config/run-axe.js:
```diff
@@ -0,0 +1,152 @@
+      });
+      Page.loadEventFired(() => {
+        Runtime.evaluate({
+          expression: AXE_JS + ';(' + runAxe.toString() + ')()',
```
Is this possible to write using a template string so it's easier to parse?

msecret commented on config/run-axe.js:
```diff
@@ -0,0 +1,152 @@
+              errorFound = true;
+            }
+          } else {
+            console.log('Unexpected result from CDP!');
```
What's CDP?

toolness replied:
Oh, it stands for Chrome DevTools Protocol... I'll expand it to just say that!

toolness commented:
Ok, I think I addressed all your issues, though I had second thoughts about ignoring sub-resource loading failures, which I [commented on above](https://github.com/18F/web-design-standards-docs/pull/344#discussion_r127084792).

toolness commented:
Hooray, thanks! 💃
